2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+
+The next rung up the ladder, "Dogs" seems like a small step. But, consider the enormous mental processing it takes for a toddler to learn to ignore the differences between the family dog Molly, a small bull terrier, and the gigantic great dane that lives down the street. Learning to abstract from "Molly" to "Dogs" takes significant mental development, but once it's learned, it's automatic, processed by the brain without the perception of thought.
+
+The linguist is concerned with the relationship between words, abstractions, and meaning. So too is the programmer. Names are your words. Your names communicate meaning and understanding. Yes, you write comments as well, but comments often stand in to explain names that aren't quite right. A well-chosen name can obviate the need for an explanatory comment.
+
+## Naming functions
+
+Communication is most effective when it spans various levels of abstraction, when it travels up and down the ladder. Just stating that you should consider abstract ideas when naming things in your code is not enough. Providing abstract naming advice also requires concrete examples in order to be convincing. Concrete examples explain abstract ideas. In order to explain the color green, you might use examples like leaves in summertime, the bottom light of a traffic signal, or the color of an emerald.
+
+Consider a requirement in your application to display a person's first and last name. Simple enough. But the name you retrieve, and all other information associated with a person, comes from an external data source that you do not own. Sometimes the data contains extra whitespace &mdash; leading, trailing, and in between. For example, the name might arrive in your application as "   Jane  Doe ". The unpredictable, extra spaces cause issues in your UI and you need to strip the whitespace out for presentation.
+
+Since you do not own the data source, you cannot fix the problem at its root. The example string above contains extra whitespace before, after, and between individual parts of the name. Simply trimming leading and trailing whitespace, as most `trim` functions do, will not be enough &mdash; it will leave extra whitespace in the middle of the string. You need to isolate the words in the name and join them with a space. `String.split` and `Enum.join`, used together, do exactly what you want.
+
+```elixir
+# In syntax familiar to most programmers
+name = "   Jane  Doe "
+list = String.split(name) # => ["Jane", "Doe"]
+Enum.join(list, " ") # => "Jane Doe"
+```
+
+Since these examples are being shown in Elixir, henceforth, they will use idiomatic Elixir syntax.[^1] Elixir's pipe operator `|>` passes the result of one operation as the first argument to the next. It affords graceful and easy-to-read function composition:
+
+```elixir
+# Returns "Jane Doe"
+"   Jane  Doe "
+|> String.split()
+|> Enum.join(" ")
+```
+
```

Comment:
I might suggest a short transition sentence to the next section that invites the reader to continue with the hypothetical implementation. Something like "Your hypothetical application needs to handle this unfortunate data gracefully."

Reply:
Changed in 745892680cded5783f9c6b545ed14fec6949b189